Make sure AuthorizedKeysCommandUser exists before setting it

artburkart opened the pull request:
Thanks for making this unique, easy-to-use solution to a very common problem.

The default behavior was to drop the AuthorizedKeysCommandUser value into the codebase regardless of whether it was supported by the installed version of `openssh-server`.  The reason I wanted to add this snippet was because I noticed `sshd` is restarted even if the config is invalid. This would only be relevant where the `openssh-server` version is `< 6.1`. I know this would only be the case on old servers, but it's a simple way to protect users exploring your tooling.

If you have any suggestions for revision, I'm all ears. :)

Commit:
AuthorizedKeysCommandUser exists before setting it

The default behavior was to drop the value into the codebase
regardless of whether it was supported

artburkart commented on install.sh:
```diff
@@ -136,7 +136,9 @@ fi
 
-./install_configure_sshd.sh
+if ! ./install_configure_sshd.sh; then
+    exit 1
```
If we failed to edit the `sshd_config` file, then we probably shouldn't continue.

michaelwittig replied:
I agree and I was hoping that my `-e` at the very top achieves this?

artburkart replied:
I thought the same. I wasn't seeing the expected behavior in my shell, but I'll test it out again and report back.

artburkart commented on install_configure_sshd.sh:
```diff
@@ -3,15 +3,20 @@
   sed -i "s:#AuthorizedKeysCommand none:AuthorizedKeysCommand ${AUTHORIZED_KEYS_COMMAND_FILE}:g" "$SSHD_CONFIG_FILE"
 else
-  if ! grep -q "AuthorizedKeysCommand ${AUTHORIZED_KEYS_COMMAND_FILE}" "$SSHD_CONFIG_FILE"; then
+  if ! grep -q "^AuthorizedKeysCommand ${AUTHORIZED_KEYS_COMMAND_FILE}" "$SSHD_CONFIG_FILE"; then
```
If the value is set, but happens to be something else, then we probably shouldn't set it.

artburkart commented on install_configure_sshd.sh:
```diff
@@ -3,15 +3,20 @@
-else
-  if ! grep -q 'AuthorizedKeysCommandUser nobody' "$SSHD_CONFIG_FILE"; then
-    echo "AuthorizedKeysCommandUser nobody" >> "$SSHD_CONFIG_FILE"
+if grep -aq 'AuthorizedKeysCommandUser' "$(which sshd)"; then
```
If `AuthorizedKeysCommandUser` isn't supported, then we shouldn't set it.

michaelwittig replied:
shouldn't the file be `$SSHD_CONFIG_FILE` ?

artburkart replied:
My thought here was that we'd check whether the string `AuthorizedKeysCommandUser` exists in the `sshd` binary to determine whether sshd supports the value. I worried checking for it in the config file could cause inconsistency if the config file is managed by something like puppet or ansible and the commented out defaults are stripped out.

I used `grep -a` because I'm searching a binary file.

What are your thoughts about this idea?

michaelwittig replied:
I see. I didn't payed attention to the `a` flag.

michaelwittig commented:
I added two notes and I can remember that one of the supported distributions did not have the config values and that's why they are added if not found. I can figure this out by running the test suite once I have your feedback on my comments.

artburkart commented:
@michaelwittig thank you for reviewing my code :)

Commit:
Uses `set -x` intead of the shebang

This way we it doesn't matter whether someone executes the install
script as `./install.sh` or `bash install.sh`, the behavior is the same.

artburkart commented:
https://github.com/widdix/aws-ec2-ssh/pull/131#discussion_r198110098

Today I learned. I was executing the `install.sh` script with `bash install.sh`, but this has the side-effect of removing the `-e` from all subshells. Check it out:

>>> Subshells spawned to execute command substitutions inherit the value of the -e option from the parent shell. When not in POSIX mode, Bash clears the -e option in such subshells.
https://www.gnu.org/software/bash/manual/bashref.html#Bash-POSIX-Mode

I was effectively removing the `-e` argument when I was executing it, which is why I was getting unexpected behavior. I definitely need to change my code back to what you had, but I am left with the question of whether it would make sense to use `set -x` on its own line instead of `bash -x` in the shebang? I pushed up a new commit. I look forward to your feedback. 😄

artburkart commented:
We are using this code on CentOS 6 and 7 boxes now :)
